Make MaltParser safe to use in parallel

pflaquerre opened the pull request:
These commits make MaltParser safe(r) to use in parallel. The paths to temporary files used to be hard-coded, which would be problematic if you're running multiple instances at the same time. There is still a potential issue if the user leaves mco to the default value, but I wasn't sure whether the user would expect the model to still be there after nltk quits or not in this case.

Changes:
- Actually use self.working_dir
- Replace hard-coded temporary file names with tempfile.NamedTemporaryFile
- Use subprocess instead of os.system to avoid quoting issues when calling java
- Error out if subprocess fails instead of blindly going forward

pflaquerre committed:
make MaltParser safe to use in a parallel setting

- Actually use self.working_dir
- Replace hard-coded temporary file names with
  tempfile.NamedTemporaryFile
- Use subprocess instead of os.system to avoid quoting issues when
  calling java
- Error out if subprocess fails

pflaquerre committed:
remove unused with_statement from malt.py

pflaquerre committed:
delete temp file even if conll_file.open fails

Just moved the code inside the try block.

heatherleaf commented:
Good things, but I don't like this:

```
def __init__(self, tagger=None, mco=None, working_dir=tempfile.gettempdir()):
    self.working_dir = working_dir
```

I think the old code is better:

```
def __init__(self, tagger=None, mco=None, working_dir=None):
    self.working_dir = tempfile.gettempdir() if working_dir is None else working_dir
```

The reason is that with your changes, the working_dir will be resolved when the module is imported. With the old code, working_dir is resolved when the MaltParser object is created. 

A minor issue, I agree.

pflaquerre committed:
initialise working_dir at __init__-time

Call gettempdir() at initialisation time instead of module load time.

pflaquerre committed:
fix malt's mco parameter documentation

As far as I could see, MaltParser does not take a full path as input for
mco models. You must specify the name of the model, which MaltParser
then uses to look for a file in ${working_dir}/${mco}.mco. It's a bit
clunky, but I couldn't find alternatives in their documentation.

pflaquerre commented:
Thanks for the comments. I've reverted that part of the code back to what it used to be. I've also updated the mco parameter documentation to better reflect what MaltParser actually does with it.

pflaquerre committed:
fix typo in malt's mco parameter documentation

Forgot a ".mco" in "${working_dir}/${mco}.mco".

stevenbird commented:
Thanks for this.
